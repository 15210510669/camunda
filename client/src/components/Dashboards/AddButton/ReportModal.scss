.ReportModal__externalSourceLink {
  font-size: 0.8em;
  color: gray;
  text-decoration: underline;
  cursor: pointer;
}

.ReportModal {
  label {
    text-align: center;
  }
  .externalInput {
    width: 100%;
  }
  .muted {
    color: gray;
  }
}
