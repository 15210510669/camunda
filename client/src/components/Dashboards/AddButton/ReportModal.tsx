/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {useEffect, useState} from 'react';
import {RouteComponentProps, withRouter} from 'react-router-dom';
import {Button} from '@carbon/react';
import {SerializedEditorState} from 'lexical';

import {
  Modal,
  Input,
  Typeahead,
  LoadingIndicator,
  Labeled,
  Form,
  CarbonTabs,
  Icon,
  TextEditor,
} from 'components';
import {getCollection, isTextReportValid, loadReports} from 'services';
import {t} from 'translation';
import {DashboardTile, GenericReport} from 'types';

interface ReportModalProps extends RouteComponentProps {
  close: () => void;
  confirm: (reportConfig: Partial<DashboardTile>) => void;
}

type TabOpen = 'report' | 'external' | 'text';

export function ReportModal({close, confirm, location}: ReportModalProps) {
  const [availableReports, setAvailableReports] = useState<GenericReport[] | null>(null);
  const [selectedReportId, setSelectedReportId] = useState<string>('');
  const [externalUrl, setExternalUrl] = useState<string>('');
  const [tabOpen, setTabOpen] = useState<TabOpen>('report');
  const [text, setText] = useState<SerializedEditorState>();

  useEffect(() => {
    (async () => {
      const collection = getCollection(location.pathname);
      const availableReports = await loadReports(collection);
      setAvailableReports(availableReports);
    })();
  }, [location.pathname]);

  const addReport = () => {
    confirm(getReportConfig(tabOpen, externalUrl, text, selectedReportId));
  };

  const getReportConfig = (
    tabOpen: TabOpen,
    externalUrl: string,
    text: SerializedEditorState | undefined,
    selectedReportId: string
  ): Partial<DashboardTile> => {
    if (tabOpen === 'external') {
      return {id: '', configuration: {external: externalUrl}, type: 'external_url'};
    } else if (tabOpen === 'text') {
      return {id: '', configuration: {text}, type: 'text'};
    }
    return {id: selectedReportId, type: 'optimize_report'};
  };

  const isExternalUrlValid = (url: string) => {
    // url has to start with https:// or http://
    return url.match(/^(https|http):\/\/.+/);
  };

  const textLength = TextEditor.getEditorStateLength(text);

  const isCurrentTabInvalid = (tabOpen: TabOpen, externalUrl: string, selectedReportId: string) => {
    const isInvalidMap = {
      report: !selectedReportId,
      external: !isExternalUrlValid(externalUrl),
      text: !isTextReportValid(textLength),
    };

    return isInvalidMap[tabOpen];
  };

  const isInvalid = isCurrentTabInvalid(tabOpen, externalUrl, selectedReportId);
  const loading = availableReports === null;
  const selectedReport = (!loading && availableReports.find(({id}) => selectedReportId === id)) || {
    id: undefined,
  };

  return (
    <Modal className="ReportModal" open onClose={close} isOverflowVisible>
      <Modal.Header>{t('dashboard.addButton.addTile')}</Modal.Header>
      <Modal.Content>
        <Form>
          <CarbonTabs<TabOpen> value={tabOpen} onChange={setTabOpen}>
            <CarbonTabs.Tab value="report" title={t('dashboard.addButton.optimizeReport')}>
              <Form.Group>
                {!loading && (
                  <Labeled label={t('dashboard.addButton.addReportLabel')}>
                    <Typeahead
                      initialValue={selectedReport.id}
                      placeholder={t('dashboard.addButton.selectReportPlaceholder')}
                      onChange={setSelectedReportId}
                      noValuesMessage={t('dashboard.addButton.noReports')}
                    >
                      <Typeahead.Option
                        key="newReport"
                        value="newReport"
                        label={`+ ${t('dashboard.addButton.newReport')}`}
                      >
                        <Icon type="plus" />
                        <b>{t('dashboard.addButton.newReport')}</b>
                      </Typeahead.Option>
                      {availableReports.map(({id, name}) => (
                        <Typeahead.Option key={id} value={id}>
                          {name}
                        </Typeahead.Option>
                      ))}
                    </Typeahead>
                  </Labeled>
                )}
                {loading && <LoadingIndicator />}
              </Form.Group>
            </CarbonTabs.Tab>
            <CarbonTabs.Tab value="external" title={t('dashboard.addButton.externalWebsite')}>
              <Form.Group>
                <Labeled label={t('dashboard.addButton.externalWebsite')}>
                  <Input
                    name="externalInput"
                    className="externalInput"
                    placeholder="https://www.example.com/widget/embed.html"
                    value={externalUrl}
                    onChange={({target: {value}}) => setExternalUrl(value)}
                  />
                </Labeled>
              </Form.Group>
            </CarbonTabs.Tab>
            <CarbonTabs.Tab value="text" title={t('dashboard.addButton.text')}>
              <Form.Group className="Labeled">
                <span className="label before">{t('dashboard.addButton.text')}</span>
                <TextEditor initialValue={text} onChange={setText} />
                <TextEditor.CharCount editorState={text} />
              </Form.Group>
            </CarbonTabs.Tab>
          </CarbonTabs>
        </Form>
      </Modal.Content>
      <Modal.Footer>
        <Button kind="secondary" onClick={close}>
          {t('common.cancel')}
        </Button>
        <Button onClick={addReport} disabled={isInvalid}>
          {t('dashboard.addButton.addTileLabel')}
        </Button>
      </Modal.Footer>
    </Modal>
  );
}

export default withRouter(ReportModal);
