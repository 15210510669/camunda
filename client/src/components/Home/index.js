/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

export {default as Home} from './Home';
export {default as Collection} from './Collection';
