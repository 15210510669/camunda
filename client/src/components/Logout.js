/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import {useEffect} from 'react';
import {withRouter} from 'react-router-dom';
import {get} from 'request';
import {withErrorHandling} from 'HOC';
import {addNotification} from 'notifications';
import {t} from 'translation';

export function Logout({mightFail, history}) {
  useEffect(() => {
    mightFail(
      get('api/authentication/logout'),
      () => {
        addNotification({text: t('navigation.logoutSuccess')});
        history.replace('/');
      },
      () => {
        addNotification({text: t('navigation.logoutFailed'), type: 'error'});
        history.replace('/');
      }
    );
  }, [mightFail, history]);

  return null;
}

export default withRouter(withErrorHandling(Logout));
