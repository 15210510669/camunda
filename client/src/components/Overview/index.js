export {default as Overview} from './Overview';
export {default as Alerts} from './Alerts';
