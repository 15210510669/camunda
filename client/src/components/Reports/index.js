export {default as Reports} from './Reports';
export {default as Report} from './Report';
export {Filter} from './filter'
