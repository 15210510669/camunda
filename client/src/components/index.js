export {Footer} from './Footer';
export {Header, LogoutButton} from './Header';
export {Home} from './Home';
export {Login} from './Login';
export {PrivateRoute} from './PrivateRoute';
export {Dashboard} from './Dashboard';
export {Dashboards} from './Dashboards';
export {Report} from './Report';
export {Reports} from './Reports';
