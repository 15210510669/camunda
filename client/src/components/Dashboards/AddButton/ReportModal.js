import React from 'react';

import {Modal, Button, Select, ControlGroup} from 'components';

import {loadReports} from '../service';

import './ReportModal.css';

export default class ReportModal extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      availableReports: null,
      selectedReportId: ''
    };
  }

  componentDidMount = async () => {
    const reports = await loadReports();

    this.setState({
      availableReports: reports
    });
  };

  selectReport = ({target: {value}}) => {
    this.setState({
      selectedReportId: value
    });
  };

  addReport = () => {
    this.props.confirm({id: this.state.selectedReportId});
  };

  render() {
    const noReports = !this.state.availableReports || this.state.availableReports.length === 0;
    const loading = this.state.availableReports === null;

    return (
      <Modal
        open
        onClose={this.props.close}
        onEnterPress={this.state.selectedReportId ? this.addReport : undefined}
      >
        <Modal.Header>Add a Report</Modal.Header>
        <Modal.Content>
          <ControlGroup layout="centered">
            <label htmlFor="ReportModal__selectReports">Select a Reportâ€¦</label>
            <Select
              disabled={noReports || loading}
              value={this.state.selectedReportId}
              onChange={this.selectReport}
              name="ReportModal__selectReports"
              className="ReportModal__selectReports"
            >
              {this.renderPleaseSelectOption(!noReports)}
              {!loading &&
                this.state.availableReports.map(report => {
                  return (
                    <Select.Option key={report.id} value={report.id}>
                      {this.truncate(report.name, 50)}
                    </Select.Option>
                  );
                })}
              {loading ? (
                <Select.Option>loading...</Select.Option>
              ) : noReports ? (
                <Select.Option>No reports created yet</Select.Option>
              ) : (
                ''
              )}
            </Select>
          </ControlGroup>
          <p className="ReportModal__externalSourceLink" onClick={this.props.gotoExternalMode}>
            Add External Source
          </p>
        </Modal.Content>
        <Modal.Actions>
          <Button onClick={this.props.close}>Cancel</Button>
          <Button
            type="primary"
            color="blue"
            onClick={this.addReport}
            disabled={!this.state.selectedReportId}
          >
            Add Report
          </Button>
        </Modal.Actions>
      </Modal>
    );
  }

  truncate = (str, index) => {
    return str.length > index ? str.substr(0, index - 1) + '\u2026' : str;
  };

  renderPleaseSelectOption = hasReports => {
    if (hasReports) {
      return (
        <Select.Option defaultValue value="">
          Please select...
        </Select.Option>
      );
    } else {
      return '';
    }
  };
}
