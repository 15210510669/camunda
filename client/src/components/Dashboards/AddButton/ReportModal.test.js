/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import {shallow} from 'enzyme';

import ReportModal from './ReportModal';
import {Button, Typeahead} from 'components';
import {loadReports} from 'services';

jest.mock('react-router-dom', () => {
  const rest = jest.requireActual('react-router-dom');
  return {
    ...rest,
    withRouter: (a) => a,
  };
});

jest.mock('services', () => {
  const rest = jest.requireActual('services');
  return {
    ...rest,
    loadReports: jest.fn().mockReturnValue([]),
  };
});

const props = {
  location: {pathname: '/dashboard/1'},
};

it('should load the available reports', () => {
  shallow(<ReportModal {...props} />);

  expect(loadReports).toHaveBeenCalled();
});

it('should load only reports in the same collection', () => {
  shallow(<ReportModal location={{pathname: '/collection/123/dashboard/1'}} />);

  expect(loadReports).toHaveBeenCalledWith('123');
});

it('should render a Typeahead element with the available reports as options', () => {
  const node = shallow(<ReportModal {...props} />);

  node.setState({
    availableReports: [
      {
        id: 'a',
        name: 'Report A',
      },
      {
        id: 'b',
        name: 'Report B',
      },
    ],
  });

  expect(node.find('Typeahead')).toMatchSnapshot();
});

it('should call the callback when adding a report', () => {
  const spy = jest.fn();
  const node = shallow(<ReportModal {...props} confirm={spy} />);

  node.setState({
    availableReports: [
      {
        id: 'a',
        name: 'Report A',
      },
      {
        id: 'b',
        name: 'Report B',
      },
    ],
    selectedReportId: 'a',
  });

  node.find('[primary]').simulate('click');

  expect(spy).toHaveBeenCalledWith({
    id: 'a',
  });
});

it('should disable typeahead if no reports created yet', async () => {
  const node = await shallow(<ReportModal {...props} />);

  expect(node.find('Typeahead')).toBeDisabled();
});

it('should show a loading message while loading available reports', () => {
  const node = shallow(<ReportModal {...props} />);

  expect(node.find('LoadingIndicator')).toExist();
});

it('should contain an Add External Source field', () => {
  const node = shallow(<ReportModal {...props} />);

  expect(node.find(Button).at(1)).toIncludeText('Add External Source');
});

it('should hide the typeahead when external mode is enabled', () => {
  const node = shallow(<ReportModal {...props} />);

  node.setState({external: true});

  expect(node.find('Typeahead')).not.toExist();
});

it('should contain a text input field if in external source mode', () => {
  const node = shallow(<ReportModal {...props} />);

  node.setState({external: true});

  expect(node.find('.externalInput')).toExist();
});

it('should  disable the submit button if the url does not start with http in external mode', () => {
  const node = shallow(<ReportModal {...props} />);

  node.setState({external: true, externalUrl: 'Dear computer, please show me a report. Thanks.'});

  expect(node.find('[primary]')).toBeDisabled();
});
