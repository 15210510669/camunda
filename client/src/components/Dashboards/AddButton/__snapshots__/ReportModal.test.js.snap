// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a Typeahead element with the available reports as options 1`] = `
<Typeahead
  disabled={false}
  noValuesMessage="No reports have been created"
  onChange={[Function]}
  onSearch={[Function]}
  placeholder="Select a Report"
>
  <ForwardRef(DropdownOption)
    key="a"
    value="a"
  >
    Report A
  </ForwardRef(DropdownOption)>
  <ForwardRef(DropdownOption)
    key="b"
    value="b"
  >
    Report B
  </ForwardRef(DropdownOption)>
</Typeahead>
`;
